### PR TITLE
feat: add app.disablePluginSandbox(mimeType)

### PR DESCRIPTION
**Pull request by miniak**
#### Description of Change
Added `app.disablePluginSandbox(mimeType)` that allows to disable sandboxing of the ppapi plugin host process, which is per-plugin and shared by multiple `BrowserWindow` instances.

Address #18516. Follow up to https://github.com/electron/electron/pull/15894#issuecomment-443781871.

/cc @nornagon 

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Added `app.disablePluginSandbox(mimeType)`, which disables sandbox for the plugin identified by `mimeType`.

### CODE REVIEW

**Comment by miniak** [comment]
@nornagon Should we maybe handle #17823 in `AtomBrowserClient::IsPluginSandboxed()` now?

**Comment by miniak** [comment]
cc @Feverqwe

**Comment by MarshallOfSound** [comment]
As a feature being targetted at `6-0-x` this late in the beta cycle it needs approval from @electron/wg-releases.  Please add this PR to the WG agenda with reasoning for it's addition

**Comment by miniak** [comment]
@MarshallOfSound added to the agenda. Regardless of whether it should be backported or not, how does the PR look to you?

File: patches/chromium/add_contentbrowserclient_ispluginsandboxdisabled_callback.patch [hunk 1]
```diff
@@ -0,0 +1,109 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Milan Burda <milan.burda@gmail.com>
+Date: Sat, 22 Jun 2019 12:23:04 +0200
+Subject: Add ContentBrowserClient::IsPluginSandboxDisabled() callback
+
+Allows the embedder to decide whether the plugin should be sandboxed.
```
Comment (nornagon) [inline patches/chromium/add_contentbrowserclient_ispluginsandboxdisabled_callback.patch]: Please upstream this patch & add a link to the chromium codereview :)

File: shell/browser/atom_browser_client.cc [hunk 2]
```diff
@@ -988,6 +989,17 @@ bool AtomBrowserClient::PreSpawnRenderer(sandbox::TargetPolicy* policy) {
 }
 #endif  // defined(OS_WIN)
 
+bool AtomBrowserClient::IsPluginSandboxDisabled(
+    const content::PepperPluginInfo& info) {
+  const auto& unsandboxed_plugins = Browser::Get()->unsandboxed_plugins();
```
Comment (nornagon) [inline shell/browser/atom_browser_client.cc]: Would it be reasonable to implement this as a callback into JS? I'm imagining an API like

```js
app.setPluginSandboxEnabledHandler((mimeType) => {
  return mimeType !== 'foo/bar'
})
```

As written, the API doesn't allow for re-enabling the sandbox after it's disabled, or for choosing which webContents are allowed to launch unsandboxed plugins. I'm not sure that it'd be likely to need such features, just asking whether we should consider them.

File: patches/chromium/add_contentbrowserclient_ispluginsandboxdisabled_callback.patch (hunk 1 above)
Comment (miniak) [reply patches/chromium/add_contentbrowserclient_ispluginsandboxdisabled_callback.patch]: I have never upstreamed stuff to Chromium so far, I need to try it for the first time. Any tips?

File: shell/browser/atom_browser_client.cc (hunk 2 above)
Comment (miniak) [reply shell/browser/atom_browser_client.cc]: I don't think you can easily map the plugin to the webContents. My thinking was that there are plugins that just don't work with sandbox at all, so we just need a way to opt-out. I don't think it's necessary to do it with more granularity.

Comment (miniak) [reply shell/browser/atom_browser_client.cc]: the JS handler on the other hand could get more information from `content::PepperPluginInfo`, including the name, path and MIME types

Comment (miniak) [reply shell/browser/atom_browser_client.cc]: @Feverqwe what do you think as the person experiencing the issue that we are trying to solve here?

**Comment by codebytere** [comment]
@miniak do you still want to merge this? it's needing rebase

File: patches/chromium/add_contentbrowserclient_ispluginsandboxdisabled_callback.patch (hunk 1 above)
Comment (nornagon) [reply patches/chromium/add_contentbrowserclient_ispluginsandboxdisabled_callback.patch]: Oops, I meant to respond to this ages ago. The best place to start is https://chromium.googlesource.com/chromium/src/+/master/docs/contributing.md

tldr: get a chromium checkout, make a branch, and run `git cl upload`.

File: shell/browser/atom_browser_client.cc (hunk 2 above)
Comment (Feverqwe) [reply shell/browser/atom_browser_client.cc]: I think that any api will be fine. setPluginSandboxEnabledHandler is more flexible, but and disablePluginSandbox looks good.

**Comment by codebytere** [comment]
Closing for now given no action and no response to the above but we can re-open should there be a path forward :)

**Comment by fp07** [comment]
Can we re-open this?
